Complex commands

psprint opened the pull request:
Basically, instead of:
```zsh
     output=$("${cmd[@]}" 2>&1)
```

it is done:
```zsh
     eval "output=$( ${cmd[@]} 2>&1)"
```

because this allows the command to include e.g. redirections or command separators like ; or ||, etc., or pipes, etc.

Actually, the eval looks more terrifying:
```zsh
  IFS=$'\n' eval "output=\$( function run {
        ${cmd[@]/(#m)*/${${${${${(M)MATCH:#(${(j:|:)~dont_quote})}:+$MATCH}}:-\"$MATCH\"}}} 2>&1 }; run )";
```

But the substitution is (as always..) simple: it quotes the command's arguments by using double quotes, so that when eval sees it, it will not skip empty strings (and also that it will not split strings). The `dont_quote` array holds things that are not to be quoted – currently the semicolon and redirections (`'[[:digit:]]>&[[:digit:]]`). The `run` function is to make the error messages look the same, i.e.:

```
run:1: command not found: a-non-existent-command
```

instead of `(eval):1: command ...`. I've checked that zunit's tests pass normally with this change.

The change allowed to create the following test-case: https://github.com/zdharma/fast-syntax-highlighting/blob/master/tests/example.zunit

psprint commented:
Any chance of the PR being merged?

psprint commented:
Maybe I could join the project? I.e. be added to the collaborators of the repo? I could help you maintain it. I have a direct need for the project (to test to f-sy-h, [declare-zsh](https://github.com/zdharma/declare-zsh) and other projects), I think that I would be an active maintainer and developer.

molovo commented:
@psprint absolutely, I'd be glad to have you on the team. Expect an invite shortly.

psprint commented:
Great, I'm really happy!

As for the PRs, do you maybe have any remarks vs them? If not, then I would
merge

On Wed, 31 Jul 2019 at 09:37, James Dinsdale <notifications@github.com>
wrote:

> @psprint <https://github.com/psprint> absolutely, I'd be glad to have you
> on the team. Expect an invite shortly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zunit-zsh/zunit/pull/104?email_source=notifications&email_token=ABOE4CDZ6MYCB2PLXUCZCCTQCE6KVA5CNFSM4IEWFTDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3GL6UY#issuecomment-516734803>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOE4CC6SPE2FMNNZB6KZALQCE6KVANCNFSM4IEWFTDA>
> .
>


-- 
Sebastian Gniazdowski
News: https://twitter.com/ZdharmaI
IRC: https://kiwiirc.com/client/chat.freenode.net:+6697/#zplugin
Blog: http://zdharma.org

molovo commented:
@psprint they both look like great additions - as long as tests are passing I'm happy for them to be merged 👍 (I really need to fix the Travis scripts)

Thanks for your hard work on this!

MichaelAquilina commented:
Could we get some documentation on this?

psprint commented:
Yes, I'll provide documentation soon.

psprint commented:
The documentation is at the `www` repository: https://github.com/zunit-zsh/www/pull/7

Could I maybe get write access so that I could merge the PRs?

molovo commented:
@psprint sorry, you should have had write access when I added you. I've fixed your permissions now so you should be able to merge them. Nice work!

psprint commented:
Great! thanks!

pon., 12 sie 2019, 14:33 użytkownik James Dinsdale <notifications@github.com>
napisał:

> @psprint <https://github.com/psprint> sorry, you should have had write
> access when I added you. I've fixed your permissions now so you should be
> able to merge them. Nice work!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zunit-zsh/zunit/pull/104?email_source=notifications&email_token=ABOE4CEBB3P3MIO4S3UXY5LQEFKCFA5CNFSM4IEWFTDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4CMISQ#issuecomment-520406090>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOE4CHRY6VD5XS2UFNWG73QEFKCFANCNFSM4IEWFTDA>
> .
>
